### PR TITLE
Use ES6, convert to ES6-style classes

### LINKED ISSUE #3: Convert to ES6
Sort out tests and node support for ES6 to ensure that we can still run tests / run the vm headless.

### PR DESCRIPTION
**Pull request by rschamp**
### Resolves

_What Github issue does this resolve (please include link)?_
Resolves #3 

### Proposed Changes

_Describe what this Pull Request does_
Updates the build system to use Babel, and updates class syntax to ES6 classes.

### Reason for Changes

_Explain why these changes should be made_
In our discussion last week we decided to settle on ES6 everywhere, including here.

### Test Coverage

_Please show how you have added tests to cover your changes_
It works!

### CODE REVIEW

File: src/blocks/scratch3_pen.js [hunk 1]
```diff
@@ -18,322 +18,337 @@ var RenderedTarget = require('../sprites/rendered-target');
...
+    }
+
     /**
-     * The runtime instantiating this block package.
-     * @type {Runtime}
+     * The default pen state, to be used when a target has no existing pen state.
+     * @type {PenState}
      */
-    this.runtime = runtime;
+    static get DEFAULT_PEN_STATE () {
```
Comment (cwillisf) [inline src/blocks/scratch3_pen.js:50]: 😎

File: src/blocks/scratch3_procedures.js [hunk 2]
```diff
@@ -1,44 +1,46 @@
...
-        var paramNames = util.getProcedureParamNames(procedureCode);
-        for (var i = 0; i < paramNames.length; i++) {
-            if (args.hasOwnProperty('input' + i)) {
-                util.pushParam(paramNames[i], args['input' + i]);
+    callNoReturn (args, util) {
+        if (!util.stackFrame.executed) {
+            const procedureCode = args.mutation.proccode;
+            const paramNames = util.getProcedureParamNames(procedureCode);
+            for (let i = 0; i < paramNames.length; i++) {
+                if (args.hasOwnProperty(`input${i}`)) {
```
Comment (cwillisf) [inline src/blocks/scratch3_procedures.js:31]: This is one of my favorite ES6 features: more readable AND (possibly) more efficient!

File: src/engine/blocks.js [hunk 3]
```diff
@@ -1,496 +1,501 @@
...
+        const tagName = (block.shadow) ? 'shadow' : 'block';
+        const xy = (block.topLevel) ?
+            ` x="${block.x}" y="${block.y}"` :
+            '';
+        let xmlString = '';
+        xmlString += `<${tagName
+            } id="${block.id}"` +
+            ` type="${block.opcode}"${
+            xy
+            }>`;
```
Comment (cwillisf) [inline src/engine/blocks.js]: The formatting got weird here RE: newlines. Looks like there's a `+` that could be eliminated, too.
I noticed a similar newline issue further down in this file, too.

File: src/import/load-costume.js [hunk 4]
```diff
@@ -13,62 +13,59 @@ var log = require('../util/log');
...
-                removeEventListeners = function () {
-                    imageElement.removeEventListener('error', onError);
-                    imageElement.removeEventListener('load', onLoad);
-                };
-                imageElement.addEventListener('error', onError);
-                imageElement.addEventListener('load', onLoad);
-                imageElement.src = costumeAsset.encodeDataURI();
-            });
-        }).then(function (imageElement) {
+        promise = promise.then(costumeAsset => new Promise((resolve, reject) => {
```
Comment (cwillisf) [inline src/import/load-costume.js]: IMO this actually became less readable... do you think a newline would help?

**Review by cwillisf — CHANGES_REQUESTED** [review]
Thanks for doing this, @rschamp!
Overall it looks great... just a couple nits.

**Comment by rschamp** [comment]
@cwillisf Agreed on your points. Also needed to merge, so merged and ES6-ified the merged code.

**Comment by rschamp** [comment]
Also just realized since we're now on Node 6, we can use ES6 syntax at the root of the project too (tests)... should I do that? I think it pretty much can be done with `eslint --fix`.

**Review by cwillisf — APPROVED** [review]

**Comment by cwillisf** [comment]
I do think we should move tests to ES6 but maybe that should be a separate PR, just due to the size and current state of this one? I don't feel strongly about that :)

**Review by thisandagain — COMMENTED** [review]
This looks good to me. I'm curious if the conversion via Babel is affecting performance. Have you run any bench tests?

**Comment by thisandagain** [comment]
Re: tests. I think it's probably good to hold off on migrating the test suite to ES6. It would good to file an issue for it though.

**Comment by rschamp** [comment]
@thisandagain Nothing super thorough, but comparing the profiles of the default project running in the current VM playground and the ES6 version, nothing stands out as being Babel-related to me.

**Pull request merged**